lark/parsers/lalr_puppet.py

```diff
@@ -1,107 +1,55 @@
-        self._value_stack = value_stack
-        self._start = start
-        self._stream = stream
-        self._set_state = set_state
-
-        self.result = None
+        self.parser_state = parser_state
+        self.lexer_state = lexer_state
 
     def feed_token(self, token):
         """Feed the parser with a token, and advance it to the next state, as if it recieved it from the lexer.
 
         Note that ``token`` has to be an instance of ``Token``.
         """
-        end_state = self.parser.parse_table.end_states[self._start]
-        state_stack = self._state_stack
-        value_stack = self._value_stack
-
-        state = state_stack[-1]
-        action, arg = self.parser.parse_table.states[state][token.type]
-        if arg == end_state:
-            raise ParseError(arg)
-
-        while action is Reduce:
-            rule = arg
-            size = len(rule.expansion)
-            if size:
-                s = value_stack[-size:]
-                del state_stack[-size:]
-                del value_stack[-size:]
-            else:
-                s = []
-
-            value = self.parser.callbacks[rule](s)
-
-            _action, new_state = self.parser.parse_table.states[state_stack[-1]][rule.origin.name]
-            assert _action is Shift
-            state_stack.append(new_state)
-            value_stack.append(value)
-
-            if state_stack[-1] == end_state:
-                self.result = value_stack[-1]
-                return self.result
-
-            state = state_stack[-1]
-            try:
-                action, arg = self.parser.parse_table.states[state][token.type]
-            except KeyError as e:
-                raise ParseError(e)
-            assert arg != end_state
-
-        assert action is Shift
-        state_stack.append(arg)
-        value_stack.append(token)
-
-    def copy(self):
+        return self.parser_state.feed_token(token)
+
+    def __copy__(self):
         """Create a new puppet with a separate state.
 
         Calls to feed_token() won't affect the old puppet, and vice-versa.
         """
         return type(self)(
             self.parser,
-            list(self._state_stack),
-            deepcopy(self._value_stack),
-            self._start,
-            self._stream,
-            self._set_state,
+            copy(self.parser_state),
+            copy(self.lexer_state),
         )
 
     def __eq__(self, other):
         if not isinstance(other, ParserPuppet):
             return False
 
-        return (
-            self._state_stack == other._state_stack and
-            self._value_stack == other._value_stack and
-            self._stream == other._stream and
-            self._start == other._start
-        )
+        return self.parser_state == other.parser_state and self.lexer_state == other.lexer_state
 
     def __hash__(self):
-        return hash((tuple(self._state_stack), self._start))
+        return hash((self.parser_state, self.lexer_state))
```

Comment:
Neither of the State classes define `__hash__` make this an (almost) useless operation. I would suggest adding `__hash__` to both classes.

Reply:
Scrap that. Remove the implementation of `__hash__`. As stated the python docs, mutable classes shouldn't implement `__hash__`